### ISSUE TITLE
Multi-language (the manual translate proposal)

### ISSUE THREAD

**Issue by marcelstoer** [mkdocs/mkdocs#774]
_&lt;edit&gt;_

### Note from project maintainers:

There are multiple aspects to supporting multiple languages in MkDocs, each of which is linked to from #211, which is the primary issue and ties all of the various aspects together. However, it is best to discuss each aspect in its related issue.

This issue is one (of two) proposals for adding support for pages in multiple languages. This issue proposes a system in which pages are manually translated, and #617 proposes a system in which pages are auto-translated using po files or the like. 

Some third party plugins exist to address this issue (see comments below). It is likely that any native solution would come from a third party plugin if/when one establishes itself as well developed and is a clear favorite among users. Whether that means the plugin would be moved in-house or it would become the "recommended" solution and remain a third-party plugin will depend upon the circumstances at the time that such a decision is made. In any event, the best way to move this forward is to volunteer your time to develop and/or test (and provide feedback to) the plugins.

_&lt;/edit&gt;_

This is related to #617 and, therefore, also to #211. However, I didn't want to add my comments to either of the two issues as to not hijack ideas discussed there. Feel free to close it if you disagree.

I'd like to share my requirements for multi-language support and ask for your input as to how to work around current limitations.
- All the docs (or doc content) must reside within the same repository/branch as the code itself. The rational is that we want to verify that each PR is complete and self-contained i.e. changes to code are properly reflected in the docs. That's the reason the [Read the Docs localization support](http://read-the-docs.readthedocs.org/en/latest/localization.html#project-with-multiple-translations) isn't really helpful, apart from the fact that it's currently only available for Sphinx-based projects. 
- Derived from the above...the localized docs must be stored _along side_ the default language docs.
- A solution that is based on externalized strings i.e. `.pot`/`.po` isn't desirable as each localization may slightly vary in how the content is represented. So, the localized version must depend on individually localized `.md` files.
- Ideally, localized versions of the docs should be free to define their own TOC. It's IMO not always useful to enforce that each localization has page X available because X might only make sense in certain languages.

So, for us a perfect solution would allow to define `mkdocs.yml#pages` for each language and to have a structure either like this

```
docs/
docs/ES
docs/FR
```

or this

```
docs/index.md
docs/index.ES.md
docs/index.FR.md
```

I'm thinking about going with the latter and writing some custom JavaScript that re-writes the links of the TOC to point to the files of the selected language. The default index page would offer a language selector and to store it in a cookie. How feasible do you think this is (still pretty new to MkDocs)?

**Comment by waylan** [comment]
@marcelstoer I have to agree that the discussion on #617 is not what I would have envisioned for multi-language docs either. Despite the fact that that discussion links to multiple different projects which successfully use `.po` files to translate the body text of documents, it certainly seems more practical to me to simply translate the documents and store them as separate Markdown files (also simpler to implement).

And now to address your proposal. I like your general approach. It is more-or-less what I would do. However, I would avoid the JavaScript magic you proposed. MkDocs is a "static site generator" so, IMO, it should generate a "static site." Therefore, I see this being the structure:

```
docs/
docs/ES
docs/FR
```

However, what resides at `docs/`? Is that the "default" language (say EN for example) or is that just a landing page which is generated from `docs/EN/index.md`? If the latter, then the `docs/` to `site/` mapping is much cleaner and there is less juggling of the URLs.

I should note that with RTD, if you go to `someproject.com/`, you are redirected to `someproject.com/en/`. Of course, that is generally a server config, and not something easy to do with a "static" site out-of-the-box (although it can be accomplished in HTML via a meta tag). My inclination is to autogenerate a landing page at `/index.html` which is built from `docs/EN/index.md` with the URLs all pointing to `/EN/somepage.html`. I suppose the "meta redirect" could be included as well, but still maintain the page as a fallback.

Moving on to navigation. First note that the TOC is the list of headings and subheadings generated from the contents of a given page. The list of pages (with their titles) is "navigation" and separate from the TOC. I assume you meant to discuss "navigation". In any event, I don't really see this as a problem. Given the above points, the entire navigation for the site would be mirrored in the directory structure. Therefore your `pages` setting might look like this (page titles omitted for simplicity):

```yaml
pages:
    - EN:
        - index.md
        - about.md
    - ES:
        - index.md
        - about.md
    - FR:
        - index.md
        - about.md
```

The trick would be in the templates to simply look at which language is being generated and only use the branch of the navigation for that language.

Of course, for user visible strings in the templates, we would still need il8n support with `.po` files etc (as per #211). However the locale would just be pulled from which language subdir we happen to be building at the moment.

To me, this seems like the most ideal and simplest solution to implement. That said, there will likely be people who want the auto-generated translations. Personally, I'm with you though. The actual documents should be manually translated as separate files alongside the default language docs.

**Comment by marcelstoer** [comment]
Thanks for your valuable feedback.

> However, I would avoid the JavaScript magic you proposed.

This was just an idea how to work around the current limitations. By no means should this be considered an enhancement proposal!
Your proposal is much more thought through, though (woah, that's a lot of th and gh :wink:).

**Comment by marcelstoer** [comment]
My JavaScript work around is still WIP but I'll eventually get there...

**Directory structure**
```
.
├── DE
│   └── index.md
├── EN
│   ├── build.md
│   ...
│   ├── index.md
│   ├── modules
│   │   └── node.md
│   ...
│   └── support.md
├── css
│   └── extra.css
├── img
│   └── favicon.png
├── index.md
└── js
    └── extra.js
```

**mkdocs.yml#pages**
```yaml
pages:
- Overview: 'index.md'
- English:
    - Home: 'EN/index.md'
    - Building the firmware: 'EN/build.md'
    ...
- Deutsch:
    - Home: 'DE/index.md'
```

**js/extra.js**
Hides the navigation tree for all but the selected language:
```javascript
var nodemcu = nodemcu || {};
(function () {
  'use strict';
  var languageCodeToNameMap = {EN: 'English', DE: 'Deutsch'};
  var defaultLanguageCode = 'EN';

  $(document).ready(function () {
    hideNavigationForAllButSelectedLanguage();
  });

  function hideNavigationForAllButSelectedLanguage() {
    // URL is like http://host/EN/build/ -> extract 'EN'
    var selectedLanguageCode = window.location.pathname.substr(1, 2);
    if (!selectedLanguageCode) {
      selectedLanguageCode = defaultLanguageCode;
    }
    var selectedLanguageName = languageCodeToNameMap[selectedLanguageCode];
    $('.subnav li span').not(':contains(' + selectedLanguageName + ')').each(function (index) {
      $(this).parent().parent().hide();
    });
  }
}());
```
**Issues**
- The "root" `index.md` currently has a brief (English) introduction followed by links to each `<LANG>/index.md` file. This is a useful entry into the documentation but in order to select a different language you need to go back to that page. I'm thinking about injecting some extra DOM elements (with JavaScript) into header, footer or navigation bar to offer a permanently visible language switcher.
- I'll most likely change the localized subfolder names from `<LANG>/` to `<lang>/` because it feels odd to see `/EN/foo/` rather than `/en/foo/`.

**Comment by waylan** [comment]
> As documented only a second level of navigation is currently supported. This hurts a lot because that one level is already "wasted" by the localization sub trees.

I believe that two level restriction has been removed in the dev version (see #6) and should be available in the next release. Might want to check it out. It could use some more testing.

**Comment by marcelstoer** [comment]
Just a heads up for those who follow this issue and/or are also interested in a JavaScript workaround until there are better alternatives:
- https://nodemcu.readthedocs.org/ is now "live".
- Hiding the navigation tree for all but the selected language as outlined above works ok. The second language I used for testing is currently not enabled. We're waiting until our translators have something ready.
- I also built a language selector that looks and behaves exactly as the one from RTD into the RTD fly-out-menu (bottom left) with JavaScript.
- I automatically generate an in-page TOC on-the-fly for all API documentation pages with JavaScript e.g. https://nodemcu.readthedocs.org/en/dev/en/modules/crypto/.
- And depending on the outcome of #752/#802 I might have to unpack the JavaScript tools again :stuck_out_tongue_winking_eye: 

I'll sooner or later wrap all of that up in a blog post.

**Comment by MagicCarpetPilot** [comment]
Hi
Is here something done in standard mkdocs? Or is it still on the backlog? Or is the solution from Marcel recommended?

It would be great to have multi-language support in the near future in standard mkdocs via an easy configuration in the yml file.

As far as I understand your ideas I think it's necessary to have a new key for the languages. That could be another template variable to have a language switcher on the site. That avoids the JavaScript magic and hidden menus and mkdocs "knows" where it finds the index files.

```
languages: 
    - EN
    - DE
    - FR
pages:
- EN:
    - Home: 'EN/index.md'
    - Building the firmware: 'EN/build.md'
    ...
- DE:
    - Home: 'DE/index.md'
    ...
- FR:
    - Home: 'FR/index.md'
    ...
```

**Comment by waylan** [comment]
It should be noted that the search documentation states:

> While search does support using multiple languages together, it is best not to add additional languages unless you really need them. Each additional language adds significant bandwidth requirements and uses more browser resources. Generally it is best to keep each instance of MkDocs to a single language.

That being the case, I expect that any solution to this proposal will need to run a separate MkDocs build for each language. We already have a similar solution in [jimporter/mike][1] which runs a separate build for each version. Presumably this would follow a similar pattern and could be implemented as a third party tool.

Finally, #211 is the official issue covering multi-language support and this issue only covers one proposal for one aspect of the feature as outlines in [this comment][2].

[1]: https://github.com/jimporter/mike
[2]: https://github.com/mkdocs/mkdocs/issues/211#issuecomment-466420384

**Comment by nathanlesage** [comment]
Hey, I really like this proposal. Let me add a quick suggestion as well for multilang support:

1. How about supporting BCP-47 instead of uppercase country codes? This way any additional operations would be much easier, as BCP-47 is the de-facto standard of localisation and therefore a lot of modules already existing are compatible with it and one wouldn't need to perform complicated translation between the built-in modules and potential externals. Besides you'd have a more finegrained support (`en-US` and `en-GB` for instance). While the finegrained support may not be immediately obvious as having a US and GB translation is indeed somewhat superfluous, this would for instance open up possibilities for countries as Switzerland to have a localisation that such companies could immediately retrieve from the language. It would make semantically sense. Additionally, domains containing uppercase letters somehow look non-aesthetical to me.
2. I would definitely opt for directories as in this case it's as easy to implement `www.domain.tld/en/` and `www.domain.tld/de/` as it would be to implement `en.domain.tld/` and `de.domain.tld/`. Using language codes inside your Markdown files (e.g. `index.en.md`) would complicate deployment a lot if you would like to use subdomains.
3. How about recursively overwritten configuration? From my perspective, it makes little sense to have the navigation for all translations (which are, as a matter of fact, still separate static pages) defined in one global file. How about having one global `mkdocs.yml` configuring anything language-independent (such as extra-css and extra-js) and then having "sub-configs" in each directory that are able to overwrite the config for each build run? Then the navigation would reside where the language files reside.
4. We would need one if/else block determining if the mkdocs installation is multi- or single-language. That is: If there are no BCP-47 -compatible subfolders, assume everything is in the docs-directory, and build as it is done currently. If there are such subfolders, build a site for each one of them, but don't have like the "default language" in the top directory as this would look weird. Instead have a default language defined in the config with a simple config-option such as `default_lang: 'en-US'`. Additionally this would maintain backward-compatibility & wouldn't force people to explicitly tell the documentation which language they're documenting in.

So a perfect directory structure could look like this:

```
├── docs               // Does not contain files, only subdirectories
|   ├── en-US          // BCP-47 compatible language tag folder
|   |   ├── index.md   // These files may be mirrored in the other directories
|   |   └── mkdocs.yml // Contains all config keys that need to be overwritten (i.e. localised)
|   ├── de-DE
|   └── fr-CH
└──mkdocs.yml          // Can also contain a navigation (to have a fallback or so), but mainly defines plugins et al.
```

**Comment by KenzaB27** [comment]
> **UPDATE 2019-02**
> 
> Everything below is now obsolete as we have dropped the multi-lang support in our project.
> 
> Just a heads up for those who follow this issue and/or are also interested in a JavaScript workaround until there are better alternatives:
> 
> * https://nodemcu.readthedocs.org/ is now "live".
> * Hiding the navigation tree for all but the selected language as outlined above works ok. The second language I used for testing is currently not enabled. We're waiting until our translators have something ready.
> * I also built a language selector that looks and behaves exactly as the one from RTD into the RTD fly-out-menu (bottom left) with JavaScript.
> * I automatically generate an in-page TOC on-the-fly for all API documentation pages with JavaScript e.g. https://nodemcu.readthedocs.org/en/dev/en/modules/crypto/.
> * And depending on the outcome of #752/#802 I might have to unpack the JavaScript tools again stuck_out_tongue_winking_eye
> 
> I'll sooner or later wrap all of that up in a blog post.

Hello, 
Can u please share your github repository with us. I've tried to implement your proposition on my project but it doesn't work at all. When I enable the extra javascript I got lots of errors on the server side  and my site only displays an error 404. 

Thanks in advance!

**Comment by marcelstoer** [comment]
@KenzaB27 you can brows https://github.com/nodemcu/nodemcu-firmware/tree/87b3ffa6bd909412bb6de6b9a056cbcf64d1bed1. That's the revision before I removed most of the hacks I outlined above. Please don't continue the discussion about my specific implementation as it is off-topic here.

**Comment by Stanzilla** [comment]
There is https://pypi.org/project/mkdocs-multilang/ anyone tried that yet?

**Comment by sim6** [comment]
mkdocs-multilang is outdated, breaks with mkdocs > 1.0

**Comment by ultrabug** [comment]
Hi everyone

Since I ended up facing the same problem as you I decided to give it a go and came up with what I hope to be an useful solution.

The mkdocs-plugin I created is called `mkdocs-static-i18n` and is available here: https://github.com/ultrabug/mkdocs-static-i18n

There's a demo docs/ + website showcasing how it works. It's simple but powerful enough to fuel my website's needs and is compatible with mkdocs-material search and mkdocs-awesome-pages plugins.

I hope that it will serve someone else. Comments & contributions are welcome of course!

Cheers

**Comment by Stanzilla** [comment]
@ultrabug this is awesome and pretty much what I would have suggested as well. Great work.

**Comment by waylan** [comment]
@ultrabug thanks for sharing. Be sure to list your plugin on the [wiki][1] so others can find it. The wiki is publicly editable so you can add it yourself.

[1]: https://github.com/mkdocs/mkdocs/wiki/MkDocs-Plugins

**Comment by ultrabug** [comment]
@waylan thanks for pointing this out, I've added it to the list!

**Comment by marcelstoer** [comment]
@ultrabug thanks for sharing your plugin, well done! It's so cool to see that someone picked up the ideas that @waylan and I were exchanging through this issue back in 2015.

I took a quick look (and submitted the first PR 😄 ). Two things I dislike:
- I need to duplicate content for the default language. That leads to more work or inconsistencies. It would be helpful if I could simply create content for languages A & B, tell the plugin that the default language be A, and the plugin would handle the rest internally (somehow). If necessary, the default content could be generated at build time.
- The name; just a detail. IMHO this is more about _localization_ than _internationalization_. Hence, it would be more appropriate to call it mkdocs-static-l10n.

**Comment by ultrabug** [comment]
@marcelstoer thanks for your kind words :)

> I need to duplicate content for the default language. That leads to more work or inconsistencies. It would be helpful if I could simply create content for languages A & B, tell the plugin that the default language be A, and the plugin would handle the rest internally (somehow). If necessary, the default content could be generated at build time.

The documentation is not clear enough I guess because you actually don't need to duplicate content for the default language. The sample docs/ structure showcases that the plugin supports every combinations, so it's actually up to you!

Either you only specify a non localized version which will be used for both the main and default language version (if it's not "fr"):
```
docs
├── index.fr.md
├── index.md
```

or you can have all files as localized ones, the main one will thus be derived from the default language one:
 
```
├── topic1
│   ├── index.en.md
│   └── index.fr.md
```

> The name; just a detail. IMHO this is more about localization than internationalization. Hence, it would be more appropriate to call it mkdocs-static-l10n.

This one is interesting because I think it depends on what we believe is the plugin's standpoint :)

[Wikipedia says](https://en.wikipedia.org/wiki/Internationalization_and_localization):

> Localization (which is potentially performed multiple times, for different locales) uses the infrastructure or flexibility provided by internationalization (which is ideally performed only once before localization, or as an integral part of ongoing development)

So in my point of view, the act of creating the per-language files is indeed localization and is the manual work that uses the infrastructure or flexibility provided by the internationalization plugin. Thus the i18n denomination of the plugin seems correct to me.

I understand point of views can diverge of course so I'm not arguing that there's an absolute truth on the matter. Just making a point for a minor detail :+1:

**Comment by waylan** [comment]
I don't want to create an argument of internationalization vs. localization. However it would be good for our users if terms were used consistently. Wikipedia [states][1]:

> Internationalization is the process of designing a software application so that it can be adapted to various languages and regions without engineering changes. Localization is the process of adapting internationalized software for a specific region or language by translating text and adding locale-specific components. Localization (which is potentially performed multiple times, for different locales) uses the infrastructure or flexibility provided by internationalization (which is ideally performed only once before localization, or as an integral part of ongoing development).

Based on that, developing the plugin is internationalization. However, the users would use the plugin for localization of their site. So, should the plugin be named by what it is or by what it does? Of course, that is up to the developer to choose. 

However, for consistency, it might make sense to use the same terms as we will for themes. That work has been started in #1778. Note that we are calling it localization as that better describes how the user will make use of the feature. That said, that is in the early stages of development and is subject to change so I can't guarantee it won't change.

Speaking of themes, @ultrabug I'm curious how your plugin would work with #1778. The current status of that PR is that it works for the `mkdocs` theme only for English and French. However, it assumes the entire site would be one language. It would be interesting to receive some feedback on what needs to happen to support multiple languages on one site. Please follow up in #1778. Just be warned that there is a merge conflict so that branch can only be used in isolation until it is resolved.

Also, for anyone interested, #1778 needs volunteers to test, provide translations, and finish development. If you would like to have translated `next/previous` links and the like in your themes, please see #1778 for details.

[1]: https://en.wikipedia.org/wiki/Internationalization_and_localization

**Comment by marcelstoer** [comment]
Thanks Waylan for all the information. In my circles we use i18n when we mean things like different date, time and number formats for different regions/countries and l10n when we turn "Hello" into "¡Hola".

> should the plugin be named by what it is or by what it does?

As most people "interacting" with a plugin are those using it rather than those writing it (I hope) I would go for the later. Hence, my surprise to see i18n. Interestingly @sim6 who wrote https://gitlab.com/mkdocs-i18n/mkdocs-i18n/ also opted for _internationalization_.

**Comment by waylan** [comment]
> In my circles we use...

That's the problem. We are all in different circles. The Wikipedia article I linked to even goes on to acknowledge that various different large corporations use these terms differently or even have created their own internal terms. In the end I expect MkDocs' documentation will likely use both terms together. In other words, we will have one section labeled "internationalization/localization". That way users familiar with either term will find it. However, that doesn't really work for a plugin name. Therefore I always go with the idea that [the one who builds the bike shed gets to paint it][1].

[1]: https://docs.freebsd.org/en_US.ISO8859-1/books/faq/misc.html#idp50108024

**Comment by ultrabug** [comment]
As my final words on this naming "argument", I'd like to say that I will follow whatever naming you want to side your project on @waylan, I have no strong opinion and it's not like my plugin is popular anyway ;)

Now to the functionality side. I checked #1778 and I did not find any `fr` related `messages.po` file so I'm not sure it's actually testable as-is unless my total lack of babel background is defeating me...

The "plan" if accepted would be to get this working with english first I guess so you'd welcome messages.po contributions to other languagues for the builtin themes only right?

> Speaking of themes, @ultrabug I'm curious how your plugin would work with #1778.

I guess I should test that then but before I do invest time in this I'd like to get a feeling of how much you'd like this feature (whatever its name) built into mkdocs @waylan please?

The merge conflicts do not seem hard to get fixed; as always with stalled PRs the main issue will be how to retain original author's credits if we move forward, but that will be a question for later I guess :)

I'm sure we can come up with something cool.

**Comment by waylan** [comment]
> before I do invest time in this I'd like to get a feeling of how much you'd like this feature (whatever its name) built into mkdocs

As you posted the above comment here, I assume you are referring to translating page content (as your plugin does) by "this" (as you would have posted to #1778 if you were referring to that feature). I recently added a comment to the [top][1] of this issue which addresses that question. It states:

> It is likely that any native solution would come from a third party plugin if/when one establishes itself as well developed and is a clear favorite among users. Whether that means the plugin would be moved in-house or it would become the "recommended" solution and remain a third-party plugin will depend upon the circumstances at the time that such a decision is made. In any event, the best way to move this forward is to volunteer your time to develop and/or test (and provide feedback to) the plugins.

I added a comment to all of the related issues as well so that first time readers of any of these issues will be aware of the current status. I intend to update these comments if/when the status changes in the future.

[1]: https://github.com/mkdocs/mkdocs/issues/774#issue-122440415

**Comment by ultrabug** [comment]
Thanks @waylan 

So I allowed myself to base a new PR #2299 from the work done in #1778 to indeed validate the POC. I added the French translation to the PR and added a comment for the newbies like me who didn't know how to get a mo file from a po file :)

I then tried what you asked for with my plugin, it works like a charm : https://github.com/ultrabug/mkdocs-static-i18n/pull/3

Combining the two result a fully localized documentation/website experience :+1: 

I hope you find this interesting

**Comment by waylan** [comment]
I am excited to announce that theme localization has just been added in #2299 and will be available in version 1.2.0 of MkDocs!

Of course, that does not address translation of page content, which this issue is addressing. But the solution there may affect how this issue proceeds.

**Comment by victorbnl** [comment]
Currently my only issue with translating is that if I'm on a french page, the site-wide nav doesn't change, and the language selector always redirect to the home page, so there are no ways to go on translated pages other than the home one. Is there a workaround to have the nav of the selected language?

**Comment by botkero** [comment]
Hello,

I would also like to make a proposal, but I cannot access to hat extent this is feasible. I also don't unterstand this discussion in detail.

**proposal**
I have noticed that every proposal suggests some kind of folder structure. Personally, it would be enough for me if I could manage the different languages directly in a *.md file.

I imagine this be a bit like a switch-case, where the default languages is x.

To show in more detail what I mean.

**exmaple.md**
```
---English
Here is my pretty little english article.
---

---German
Hier ist mein schöner kleiner Artikel.
---
```
(I had to use ---, otherwise the formatting breaks in github. I am completely open as to which characters to chose.)

I am well aware that a single file can became quite large, but my impression is that this is a workaround anyway and not one that aims to support >24 languages at the time. (i think most will not need more than 3 languages).

Apart from that, this is only a suggestion. Please do not be afraid to say that you find this solution insufficient or inappropriate for the project.

**Comment by nazarepiedady** [comment]
Greetings to everyone!
Is the design that will be used for the translation structure already decided?

**Comment by ultrabug** [comment]
@nazarepiedady hi

There's no definitive answer here but the Wiki lists available solutions for you to pick : https://github.com/mkdocs/mkdocs/wiki/MkDocs-Plugins#internationalization--localization

Cheers

**Comment by ultrabug** [comment]
@oprypin I wonder if we should close this issue since it's obvious and documented now that MkDocs does support its own theme translations while content is up to plugins. What do you think?

If you agree, we should close #211 and #617 at the same time.

**Comment by oprypin** [comment]
This proposal (at least the first post) is about translating site content, rather than themes. Also using multiple languages at once. MkDocs doesn't do that

**Comment by ultrabug** [comment]
> MkDocs doesn't do that

I'm under the impression that's exactly what I said and hence proposing to close those issues.

**Comment by marcelstoer** [comment]
As the original author of this issue - gee, how time flies - I support @ultrabug's  proposal of closing this one and the other two related issues. While closing (I leave that up to the maintainers) I also propose to leave a comment stating that https://pypi.org/project/mkdocs-static-i18n is the closest thing supporting exactly what was asked here.

**Comment by ultrabug** [comment]
Thanks @marcelstoer for your approval and kind words. I'm happy I could be of help.

I closed the other issues with the given message:

i18n is now supported by mkdocs themes and there are [plugins for content localization available](https://github.com/mkdocs/mkdocs/wiki/MkDocs-Plugins#internationalization--localization) that should be considered (mkdocs-static-i18n is most [mature at the time of writing this comment](https://github.com/mkdocs/mkdocs/issues/774#issuecomment-1207802122) - disclaimer: I'm the author of the mkdocs-static-i18n plugin).
